datafusion/optimizer/src/optimizer.rs

```diff
@@ -476,6 +520,38 @@ pub(crate) fn assert_schema_is_the_same(
     }
 }
 
+/// Returns an error if plan, and subplans, do not have unique fields.
+///
+/// This invariant is subject to change.
+/// refer: <https://github.com/apache/datafusion/issues/13525#issuecomment-2494046463>
+fn assert_unique_field_names(plan: &LogicalPlan) -> Result<()> {
+    plan.schema().check_names()?;
```

Comment:
Is `check_names` also called whenever creating new DFSchema?

Reply:
Yes, on every creation. But not after [every merge](https://github.com/apache/datafusion/blob/bd2c975df88c9e95f2f6b5f2abed5638904e2ac0/datafusion/common/src/dfschema.rs#L294). Which should be ok if no bug is introduced in the merge -- altho I would prefer to add the check there.